### PR TITLE
Add `cudf::stable_sort_by_key`

### PR DESCRIPTION
**Pull request by PointKernel**
This PR adds a new `stable_sort_by_key` API into libcudf. The new API is helpful to simplify Cython/JNI bindings of `drop_duplicates` (https://github.com/rapidsai/cudf/pull/10370).

### CODE REVIEW

File: cpp/src/sort/sort.cu
Comment (davidwendt) [inline cpp/src/sort/sort.cu:48]: This change seems wrong to me. The `sorted_order` is only a temporary variable here and so should not use the `mr` to create it.

Comment (PointKernel) [reply cpp/src/sort/sort.cu:48]: Thanks for clarifying. Corrected.

File: cpp/tests/sort/stable_sort_tests.cpp
Comment (davidwendt) [inline cpp/tests/sort/stable_sort_tests.cpp:29]: I think most of these includes are not required here perhaps?
```suggestion
#include <cudf/sorting.hpp>
#include <cudf/table/table_view.hpp>
```

**Review by davidwendt — CHANGES_REQUESTED** [review]

File: cpp/tests/sort/stable_sort_tests.cpp
Comment (davidwendt) [inline cpp/tests/sort/stable_sort_tests.cpp:49]: The test types here do not include `bool` but the test code below includes special logic for handling `bool` types.

```suggestion
using TestTypes = cudf::test::Concat<cudf::test::NumericTypes,  // includes integers, floats, and bool
                                     cudf::test::ChronoTypes>; // includes timestamp and duration types
```

Are there test cases for `FixedPointTypes`? Maybe those are not sortable?

Comment (PointKernel) [reply cpp/tests/sort/stable_sort_tests.cpp]: Fixed point tests added.

File: cpp/tests/sort/stable_sort_tests.cpp
Comment (PointKernel) [reply cpp/tests/sort/stable_sort_tests.cpp]: I've removed some of them and keep `<cudf/types.hpp>` and `<cudf/copying.hpp>` to avoid transitive inclusions.

**Review by davidwendt — APPROVED** [review]

**Comment by PointKernel** [comment]
@gpucibot merge

**Comment by davidwendt** [comment]
@PointKernel you should be getting two C++ approvals for libcudf PRs before merging.

**Comment by PointKernel** [comment]
> @PointKernel you should be getting two C++ approvals for libcudf PRs before merging.

Thanks for the reminder. I realized this right after merging.

**Pull request merged**
